### ISSUE TITLE
Question: support for new JSEP plugin system (and new node types)

### ISSUE THREAD

**Issue by 6utt3rfly** [donmccurdy/expression-eval#61]
I've been helping to update jsep [(PR)](https://github.com/EricSmekens/jsep/pull/152) to allow for plugins to extend it further. It's not quite released yet, but it adds the option of numerous supplied plugins, so that it can support things like arrow-function, template-literals, objects, etc.

I wonder if [expression-eval](https://github.com/donmccurdy/expression-eval) could be made extensible to support new node types? I was thinking that one option might be converting the switch statement to a map of `{ [key: keyof ExpressionType]: (node: Expression, context?: object) => any` perhaps? Or perhaps another way to allow users to add evaluators for arbitrary node types?

I could also help with support for some of the new node types if wanted?

**Comment by donmccurdy** [comment]
This is a great idea, and I think it would be good for the project. But it has been quite a while since I've been actively working on this repository, and so it's not work than I can take on right now. Are you interested in writing PRs for this in expression-eval?

**Comment by 6utt3rfly** [comment]
Yeah I'd be interested, although I'm busy this week. Otherwise I was thinking of forking or writing something from scratch before I thought I'd ask. Did you have any ideas about how you would like it to look when done?

**Comment by donmccurdy** [comment]
I don't mind if you'd prefer to fork or write from scratch, this is just a ~300 LOC repository so there's not much architecture to speak of. :)

Perhaps the most obvious API would be similar to the current `addUnaryOp` and `addBinaryOp` methods, creating something like `addExpressionType(...)`, but I don't have much preference.

**Comment by 6utt3rfly** [comment]
I've nearly ready to create a PR, without trying to change it too much (although I'm guessing that after incremental changes a diff probably looks like a lot!). But the one thing that keeps bugging me is the sync vs async methods. They are nearly carbon copies of each other and I keep feeling like there has to be a better way vs duplicating nearly half the file.

One thought I had was a pre-build step that would read the file and with a bit of processing, would generate async copies. This might work now, but could take a bit of work to keep it from breaking...

I looked into some modules like deasync but most solutions seem to for node apps only.

Any thoughts on the code duplication between sync/async flows?

**Comment by 6utt3rfly** [comment]
> Any thoughts on the code duplication between sync/async flows?

I came up with a solution that I'm pretty happy with, which avoids all the code duplication (especially as I added more node types). I've provided a draft PR (#63) that can hopefully be used :-)

**Comment by donmccurdy** [comment]
Sorry for the delay in responding here. I will be archiving this repository and deprecating the npm package. See: https://github.com/donmccurdy/expression-eval/commit/0a8d3f3038556244f5163d61afcf53950bf6e71a.
